src/pybind/mgr/cephadm/services/monitoring.py

```diff
@@ -36,8 +36,13 @@ def generate_config(self, daemon_spec: CephadmDaemonDeploySpec) -> Tuple[Dict[st
             prom_services.append(build_url(scheme='http', host=addr, port=port))
 
             deps.append(dd.name())
+
+        loki_services = []  # type: List[str]
+        for dd in self.mgr.cache.get_daemons_by_service('mgr'):
+            addr = self.mgr.inventory.get_addr(dd.hostname)
+            loki_services.append(build_url(scheme='http', host=addr, port=3100))
         grafana_data_sources = self.mgr.template.render(
-            'services/grafana/ceph-dashboard.yml.j2', {'hosts': prom_services})
+            'services/grafana/ceph-dashboard.yml.j2', {'hosts': prom_services, 'loki_host': loki_services[0]})
```

Comment:
A bit confused by this bit here. We're setting up urls for loki services to point to IPs for whatever hosts the mgr daemons are on? Does this imply loki is meant to run on the same hosts as the mgr daemons? Does this still work if there is no loki service altogether? We're not checking for any of that here, just getting hosts with mgr daemons. Also, why are we building loki_services list with urls for each mgr daemon host if we only use the first one in the list?

Reply:
**Does this imply loki is meant to run on the same hosts as the mgr daemons?** - Yes, it is supposed to run on the same hosts as the mgr daemons.